rsconnect/bundle.py

```diff
@@ -746,6 +952,84 @@ def make_html_bundle(
     return bundle_file
 
 
+def make_voila_bundle(
+    path: str,
+    entrypoint: str,
+    extra_files: typing.List[str],
+    excludes: typing.List[str],
+    force_generate: bool,
+    environment: Environment,
+    image: str = None,
+) -> typing.IO[bytes]:
+    """
+    Create an voila bundle, given a path and/or entrypoint.
+
+    The bundle contains a manifest.json file created for the given notebook entrypoint file.
+    If the related environment file (requirements.txt) doesn't
+    exist (or force_generate is set to True), the environment file will also be written.
+
+    :param path: the file, or the directory containing the files to deploy.
+    :param entry_point: the main entry point.
+    :param extra_files: a sequence of any extra files to include in the bundle.
+    :param excludes: a sequence of glob patterns that will exclude matched files.
+    :param force_generate: bool indicating whether to force generate manifest and related environment files.
+    :param image: the optional docker image to be specified for off-host execution. Default = None.
+    :return: a file-like object containing the bundle tarball.
+    """
+    extra_files = list(extra_files) if extra_files else []
+    entrypoint_candidates = infer_entrypoint_candidates(path=abspath(path), mimetype="text/ipynb")
+    multi_notebook_mode = False
+    deploy_dir = abspath(path)
+
+    if len(entrypoint_candidates) <= 0:
+        if entrypoint is None:
+            raise RSConnectException("No valid entrypoint found.")
+        deploy_dir = dirname(entrypoint)
+    elif len(entrypoint_candidates) == 1:
+        entrypoint = entrypoint or entrypoint_candidates[0]
+        deploy_dir = dirname(entrypoint)
+    else:  # len(entrypoint_candidates) > 1:
+        if entrypoint is None:
+            raise RSConnectException(
+                """
+                Unable to infer entrypoint from multiple candidates.
+                Multi-notebook deployments need to be specified with the following:
+                1) An empty string entrypoint, i.e. ""
+                2) A directory as the path
+                """
+            )
+        elif entrypoint == "":
+            multi_notebook_mode = True
+            deploy_dir = entrypoint = abspath(path)
+        else:
+            deploy_dir = dirname(entrypoint)
+
+    voila_json_path = join(deploy_dir, "voila.json")
+    if isfile(voila_json_path):
+        extra_files.append(voila_json_path)
+
+    manifest = create_voila_manifest(**locals())
+    if manifest.data.get("files") is None:
+        raise RSConnectException("No valid files were found for the manifest.")
+
+    manifest_path = join(deploy_dir, "manifest.json")
+    manifest_flattened_copy_data = manifest.flattened_copy.data
+    if multi_notebook_mode and "metadata" in manifest_flattened_copy_data:
+        manifest_flattened_copy_data["metadata"]["entrypoint"] = ""
+    write_manifest_json(manifest_path, manifest_flattened_copy_data)
```

Comment:
This creates manifest.json in the deployment directory. We should be creating it in the Bundle but not modifying the contents of the current directory.

Reply:
fixed by [6ba4a38](https://github.com/rstudio/rsconnect-python/pull/277/commits/6ba4a382f6bbe5ea6cc32593e0ee5f026884c86e)